core/spec/models/spree/promotion_code/batch_builder_spec.rb

```diff
@@ -0,0 +1,79 @@
+require "spec_helper"
+
+describe Spree::PromotionCode::BatchBuilder do
+  let(:promotion) { create(:promotion) }
+  let(:base_code) { "abc" }
+  let(:promotion_batch) do
+    Spree::PromotionCodeBatch.create!(
+      promotion_id: promotion.id,
+      base_code: base_code,
+      number_of_codes: number_of_codes
+    )
+  end
+  let(:delivery) { double }
+  subject { described_class.new promotion_batch }
+
+  before do
+    allow(delivery).to receive(:deliver_now)
+
+    allow(Spree::PromotionCodeBatchMailer)
+      .to receive(:promotion_code_batch_finished)
+      .and_return(delivery)
+
+    allow(Spree::PromotionCodeBatchMailer)
+      .to receive(:promotion_code_batch_errored)
+      .and_return(delivery)
+  end
+
+  describe "#build_promotion_codes" do
+    context "with a failed build" do
+      let(:number_of_codes) { 10 }
+
+      before do
+        allow(subject).to receive(:generate_random_codes).and_raise "Error"
+        expect { subject.build_promotion_codes }.to raise_error RuntimeError
+      end
+
+      it "updated the error on promotion batch" do
+        expect(promotion_batch.reload.error).to eq("#<RuntimeError: Error>")
+      end
+
+      it "sends an email email" do
+        expect(Spree::PromotionCodeBatchMailer)
+          .to have_received(:promotion_code_batch_errored)
+        expect(delivery).to have_received(:deliver_now)
+      end
+    end
+    context "with a successful build" do
+      let(:number_of_codes) { 2 }
+
+      before do
+        subject.build_promotion_codes
+      end
+
+      it "update the promotion codes count for the batch" do
+        expect(promotion_batch.promotion_codes.count).to eq(number_of_codes)
+      end
+
+      it "builds the correct number of codes" do
+        expect(subject.promotion.codes.size).to eq(number_of_codes)
+      end
+
+      it "builds codes with distinct values" do
+        expect(subject.promotion.codes.map(&:value).uniq.size).to eq(number_of_codes)
+      end
+
+      it "builds codes with the same base prefix" do
+        values = subject.promotion.codes.map(&:value)
+        expect(values.all? { |val| val.starts_with?("#{base_code}_") }).to be true
+      end
+
+
+      it "sends a success email" do
+        expect(Spree::PromotionCodeBatchMailer)
+          .to have_received(:promotion_code_batch_finished)
+        expect(delivery).to have_received(:deliver_now)
```

Comment:
Please expect `ActionMailer::Base.deliveries` to be present instead of stubbing implementation details. Thanks